src/dbnode/storage/namespace.go

```diff
@@ -924,20 +926,90 @@ func (n *dbNamespace) Flush(
 		}
 
 		// skip flushing if the shard has already flushed data for the `blockStart`
-		if s := shard.FlushState(blockStart); s.Status == fileOpSuccess {
+		if s := shard.FlushState(blockStart); s.WarmStatus == fileOpSuccess {
 			continue
 		}
 		// NB(xichen): we still want to proceed if a shard fails to flush its data.
 		// Probably want to emit a counter here, but for now just log it.
-		if err := shard.Flush(blockStart, flushPersist); err != nil {
+		if err := shard.WarmFlush(blockStart, flushPersist); err != nil {
 			detailedErr := fmt.Errorf("shard %d failed to flush data: %v",
 				shard.ID(), err)
 			multiErr = multiErr.Add(detailedErr)
 		}
 	}
 
 	res := multiErr.FinalError()
-	n.metrics.flush.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	n.metrics.flushWarmData.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
+
+// idAndBlockStart is the composite key for the genny map used to keep track of
+// dirty series that need to be ColdFlushed.
+type idAndBlockStart struct {
+	id         ident.ID
+	blockStart xtime.UnixNano
+}
+
+type coldFlushReuseableResources struct {
+	// dirtySeries is a map from a composite key of <series ID, block start>
+	// to an element in a list in the dirtySeriesToWrite map. This composite key
+	// is deliberately made so that this map stays one level deep, making it
+	// easier to share between shard loops, minimizing the need for allocations.
+	dirtySeries *dirtySeriesMap
+	// dirtySeriesToWrite is a map from block start to a list of dirty series
+	// that have yet to be written to disk.
+	dirtySeriesToWrite map[xtime.UnixNano]*idList
```

Comment:
Can you remind me why we need this list? Seem to me like it could just be `map[xtime.UnixNano]*dirtySeriesMap` since a map can be used for both checking and iteration. Maybe you'll need to change `idAndBlockStart` to have a slice of blockStart but still seems fine.

I'm sure I'm missing something though

Reply:
`dirtySeriesMap` needs to have both series ID and block start because we need to know if a series ID for a block is dirty. We also need to iterate through the dirty series by block start and we can't do that efficiently with that key.

It would be cleanest to just use `map[xtime.UnixNano]map[series ID]bool`, but we didn't want a multi-level map to avoid having to keep allocating when inserting into the series ID map.

Reply:
Why would the mutli-level map require more allocations? the seriesID map could be pooled as well. Plus you can use the unsafe feature so that the series IDs dont need to be allocated since they're immutable anyways

Reply:
Spoke offline - having a map pool shared between cold flushes would work, but it's likely that maps would have varying degrees of usage, so it's possible that a small map is taken to do a lot of writes, which would not be optimal. Having a flat data structure makes pooling more uniform.

However, this means that we are creating lots of small objects (list elements) every cold flush. Let's keep this implementation for now and see if it becomes an issue.